File: progreswheelview/src/main/java/com/app/progresviews/ProgressLine.java
package com.app.progresviews;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;

import com.app.progreswheelview.R;

import java.text.DecimalFormat;

/**
 * Created by Zeki Guler on 01,February,2016
 * ©2015 Appscore. All Rights Reserved
 */
public class ProgressLine extends View {

    //Sizes (with defaults)
    private float  mBarWidth = 24f;
    private float mValueTextSize = 48;
    private float mDefTextSize = 24;
    private int layoutHeight = 0;
    private int layoutWidth = 0;

    //Colors (with defaults)
    private int    mProgressColor   = 0x61b50CCC;//Color.GREEN;
    private int    mUnderLineColor  = Color.GRAY;

    //Rectangles
    private RectF mUnderLineBounds = new RectF();
    private RectF mProgressBounds  = new RectF();

    //Painy
    private TextPaint mValueTextPaint = new TextPaint();
    private TextPaint mDefTextPaint = new TextPaint();
    private Paint mProgressLinePaint = new Paint();
    private Paint mUnderLinePaint = new Paint();


    //Padding (with defaults)
    private int marginTopBars       = 30;
    private int paddingBetweenText = 20;
    private int defaultPaddingLeft  = 30;
    private int defaultPaddingRight = 30;
    private int widthPaddingBetweenBars = 3;

    //PAdding (with defaults)
    private int leftPadding;
    private int rithPadding;



    private String mValueText = "6,0090";
    private String mDefText = "daily steps";
    private float mValueTextWidth;

    private int mPercentage = 70;
    private int mBarLenght;
    private float mScale = 1;

    private DecimalFormat mformatter = new DecimalFormat("#,###,###");

    public ProgressLine(Context context) {
        super(context);
        init(null, 0);
    }

    public ProgressLine(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public ProgressLine(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        // Load attributes
        final TypedArray a = getContext().obtainStyledAttributes(
                attrs, R.styleable.ProgressLine, defStyle, 0);

        if(a.hasValue(R.styleable.ProgressLine_value))
            mValueText  =    a.getString(R.styleable.ProgressLine_value);
        if(a.hasValue(R.styleable.ProgressLine_definition))
            mDefText = a.getString(R.styleable.ProgressLine_definition);

        mBarWidth       = a.getDimension(R.styleable.ProgressLine_lineBarWidth, mBarWidth);
        mProgressColor  = a.getColor(R.styleable.ProgressLine_lineProgressColor, mProgressColor);
        mDefTextSize    = a.getDimension(R.styleable.ProgressLine_lineDefTextSize, mDefTextSize);
        mValueTextSize  = a.getDimension(R.styleable.ProgressLine_valueDefTextSize, mValueTextSize);
        mPercentage     = a.getInt(R.styleable.ProgressLine_valuePercentage, mPercentage);

        calculateBarScale();

        a.recycle();

        // Update TextPaint and text measurements from attributes
        invalidate();
    }

    private void calculateBarScale() {
        mScale = mPercentage > 100 ?  1 :  mPercentage * 0.01f;
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        layoutWidth = w;
        layoutHeight = h;
        setupBounds();
        setupPaints();

        invalidate();
    }

    private void setupPaints() {
        mValueTextPaint.setColor(Color.BLACK);
        mValueTextPaint.setFlags(Paint.ANTI_ALIAS_FLAG);

        mDefTextPaint.setColor(Color.BLACK);
        mDefTextPaint.setFlags(Paint.ANTI_ALIAS_FLAG);


        mProgressLinePaint.setColor(mProgressColor);
        mProgressLinePaint.setAntiAlias(true);
        mProgressLinePaint.setStyle(Paint.Style.FILL);
        mProgressLinePaint.setStrokeWidth(mBarWidth);

        mUnderLinePaint.setColor(mUnderLineColor);
        mUnderLinePaint.setAntiAlias(true);
        mUnderLinePaint.setStyle(Paint.Style.FILL);
        mUnderLinePaint.setStrokeWidth(3);


        mBarLenght = this.getWidth() - defaultPaddingLeft - defaultPaddingRight - getPaddingLeft() - getPaddingRight();

    }

    private void setupBounds() {

        // Count number text
        mValueTextPaint.setTextSize(mValueTextSize);
        mValueTextWidth = mValueTextPaint.measureText(mValueText);


        // Definition text
        mDefTextPaint.setTextSize(mDefTextSize);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        float startX_V = getPaddingLeft() + defaultPaddingLeft;
        canvas.drawText(mValueText,
                startX_V,
                this.getHeight() / 2 ,
                mValueTextPaint
        );

        float start_D = startX_V + mValueTextWidth + paddingBetweenText;
        canvas.drawText(mDefText,
                start_D,
                this.getHeight() / 2 ,
                mDefTextPaint
        );

        float startX_P    = getPaddingLeft() + defaultPaddingLeft;
        float endX_P      = startX_P + mBarLenght * mScale;
        canvas.drawLine(
                startX_P,
                this.getHeight() / 2 + marginTopBars,
                endX_P,
                this.getHeight() / 2 + marginTopBars,
                mProgressLinePaint);

        float startX_U = endX_P + widthPaddingBetweenBars;
        float endX_U   = startX_U + mBarLenght * (1 - mScale);
        canvas.drawLine(
                startX_U,
                this.getHeight() / 2 + mBarWidth / 2 + marginTopBars,
                endX_U,
                this.getHeight() / 2  + mBarWidth / 2 + marginTopBars,
                mUnderLinePaint);
    }

    public void setmDefText(String mDefText) {
        this.mDefText = mDefText;
        invalidate();
    }

    public void setmValueText(int value) {
        this.mValueText = mformatter.format(value);
        invalidate();
    }

    public void setmPercentage(int mPercentage) {
        this.mPercentage = mPercentage;
        calculateBarScale();
        invalidate();
    }
}

File: progreswheelview/src/main/java/com/app/progresviews/ProgressLine.kt
package com.app.progresviews

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.text.TextPaint
import android.util.AttributeSet
import android.view.View

import com.app.progreswheelview.R

import java.text.DecimalFormat

/**
 * Created by Zeki Guler on 01,February,2016
 * ©2015 Appscore. All Rights Reserved
 */
class ProgressLine : View {

    //Sizes (with defaults)
    private var mBarWidth = 24f
    private var mUnderLineSize = 5f
    private var mValueTextSize = 48f
    private var mDefTextSize = 24f
    private var layoutHeight = 0
    private var layoutWidth = 0

    //Colors (with defaults)
    private var mProgressColor = 0x61b50CCC//Color.GREEN;
    private var mUnderLineColor = Color.GRAY

    //Rectangles
    private val mUnderLineBounds = RectF()
    private val mProgressBounds = RectF()

    //Painy
    private val mValueTextPaint = TextPaint()
    private val mDefTextPaint = TextPaint()
    private val mProgressLinePaint = Paint()
    private val mUnderLinePaint = Paint()


    //Padding (with defaults)
    private val marginTopBars = 30
    private val paddingBetweenText = 20
    private val defaultPaddingLeft = 30
    private val defaultPaddingRight = 30
    private val widthPaddingBetweenBars = 9

    //PAdding (with defaults)
    private val leftPadding: Int = 0
    private val rithPadding: Int = 0


    private var mValueText: String? = "6,0090"
    private var mDefText: String? = "daily steps"
    private var mValueTextWidth: Float = 0.toFloat()

    private var mPercentage = 70
    private var mBarLenght: Int = 0
    private var mScale = 1f

    private val mformatter = DecimalFormat("#,###,###")

    constructor(context: Context) : super(context) {
        init(null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        // Load attributes
        val a = context.obtainStyledAttributes(
                attrs, R.styleable.ProgressLine, defStyle, 0)

        if (a.hasValue(R.styleable.ProgressLine_value))
            mValueText = a.getString(R.styleable.ProgressLine_value)
        if (a.hasValue(R.styleable.ProgressLine_definition))
            mDefText = a.getString(R.styleable.ProgressLine_definition)

        mBarWidth = a.getDimension(R.styleable.ProgressLine_lineBarWidth, mBarWidth)
        mProgressColor = a.getColor(R.styleable.ProgressLine_lineProgressColor, mProgressColor)
        mDefTextSize = a.getDimension(R.styleable.ProgressLine_lineDefTextSize, mDefTextSize)
        mValueTextSize = a.getDimension(R.styleable.ProgressLine_valueDefTextSize, mValueTextSize)
        mPercentage = a.getInt(R.styleable.ProgressLine_valuePercentage, mPercentage)
        mUnderLineSize = a.getDimension(R.styleable.ProgressLine_underLineSize, mUnderLineSize)
        mUnderLineColor = a.getColor(R.styleable.ProgressLine_underLineColor, mUnderLineColor)

        calculateBarScale()

        a.recycle()

        // Update TextPaint and text measurements from attributes
        invalidate()
    }

    private fun calculateBarScale() {
        mScale = if (mPercentage > 100) 1f else mPercentage * 0.01f
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        layoutWidth = w
        layoutHeight = h
        setupBounds()
        setupPaints()

        invalidate()
    }

    private fun setupPaints() {
        mValueTextPaint.color = Color.BLACK
        mValueTextPaint.flags = Paint.ANTI_ALIAS_FLAG

        mDefTextPaint.color = Color.BLACK
        mDefTextPaint.flags = Paint.ANTI_ALIAS_FLAG


        mProgressLinePaint.color = mProgressColor
        mProgressLinePaint.isAntiAlias = true
        mProgressLinePaint.style = Paint.Style.FILL
        mProgressLinePaint.strokeWidth = mBarWidth

        mUnderLinePaint.color = mUnderLineColor
        mUnderLinePaint.isAntiAlias = true
        mUnderLinePaint.style = Paint.Style.FILL
        mUnderLinePaint.strokeWidth = mUnderLineSize


        mBarLenght = this.width - defaultPaddingLeft - defaultPaddingRight - paddingLeft - paddingRight

    }

    private fun setupBounds() {

        // Count number text
        mValueTextPaint.textSize = mValueTextSize
        mValueTextWidth = mValueTextPaint.measureText(mValueText)


        // Definition text
        mDefTextPaint.textSize = mDefTextSize
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val startX_V = (paddingLeft + defaultPaddingLeft).toFloat()
        canvas.drawText(mValueText!!,
                startX_V,
                (this.height / 2).toFloat(),
                mValueTextPaint
        )

        val start_D = startX_V + mValueTextWidth + paddingBetweenText.toFloat()
        canvas.drawText(mDefText!!,
                start_D,
                (this.height / 2).toFloat(),
                mDefTextPaint
        )

        val startX_P = (paddingLeft + defaultPaddingLeft).toFloat()
        val endX_P = startX_P + mBarLenght * mScale
        canvas.drawLine(
                startX_P,
                (this.height / 2 + marginTopBars).toFloat(),
                endX_P,
                (this.height / 2 + marginTopBars).toFloat(),
                mProgressLinePaint)

        val startXU = endX_P + widthPaddingBetweenBars
        val endXU = startXU + mBarLenght * (1 - mScale)
        canvas.drawLine(
                startXU,
                (this.height / 2).toFloat() + mBarWidth / 2 + marginTopBars.toFloat(),
                endXU,
                (this.height / 2).toFloat() + mBarWidth / 2 + marginTopBars.toFloat(),
                mUnderLinePaint)
    }

    fun setmDefText(mDefText: String) {
        this.mDefText = mDefText
        invalidate()
    }

    fun setmValueText(value: Int) {
        this.mValueText = mformatter.format(value.toLong())
        setupBounds()
        setupPaints()
        invalidate()
    }

    fun setmValueText(value: String) {
        this.mValueText = value
        setupBounds()
        setupPaints()
        invalidate()
    }

    fun setmPercentage(percetage: Int) {

        val diff = percetage - mPercentage
        ValueAnimator()
        val valueAnimator = ValueAnimator
                .ofInt(mPercentage, mPercentage + diff)
                .setDuration(1000)
        valueAnimator.addUpdateListener { animation ->
            mPercentage = animation.animatedValue as Int
            setupBounds()
            setupPaints()
            calculateBarScale()
            invalidate()
        }
        valueAnimator.start()
    }
}
